x-pack/plugins/security/server/authorization/authorization_service.tsx

```diff
@@ -154,6 +165,36 @@ export class AuthorizationService {
     initAPIAuthorization(http, authz, loggers.get('api-authorization'));
     initAppAuthorization(http, authz, loggers.get('app-authorization'), features);
 
+    http.registerOnPreResponse((request, preResponse, toolkit) => {
+      if (preResponse.statusCode === 403 && canRedirectRequest(request)) {
+        const basePath = http.basePath.get(request);
+        const next = `${basePath}${request.url.path}`;
+        const regularBundlePath = `${basePath}/${buildNumber}/bundles`;
+
+        const logoutUrl = http.basePath.prepend(
+          `/api/security/logout?${querystring.stringify({ next })}`
+        );
+        const styleSheetPaths = [
+          `${regularBundlePath}/kbn-ui-shared-deps/${UiSharedDeps.baseCssDistFilename}`,
+          `${regularBundlePath}/kbn-ui-shared-deps/${UiSharedDeps.lightCssDistFilename}`,
+          `${basePath}/node_modules/@kbn/ui-framework/dist/kui_light.css`,
+          `${basePath}/ui/legacy_light_theme.css`,
+        ];
+        const uiPublicUrl = `${basePath}/ui`;
+
+        const body = renderToStaticMarkup(
+          <ResetSessionPage
+            logoutUrl={logoutUrl}
+            styleSheetPaths={styleSheetPaths}
+            uiPublicUrl={uiPublicUrl}
+          />
+        );
+
+        return toolkit.render({ body });
```

Comment:
It looks like we are missing the CSP headers in the response. This leaves the page open to CSP bypass, which isn't ideal. While the page itself is rather trivial, we really should make sure that it's protected by our CSP.

You can get the CSP header via `http.csp.header`:

 ```suggestion
         return toolkit.render({ body, headers: { 'Content-Security-Policy': http.csp.header } });
```

One side effect of this change is that your inline JS to "go back" will no longer fire since that violates the CSP:

> Refused to run the JavaScript URL because it violates the following Content Security Policy directive: "script-src 'unsafe-eval' 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-...'), or a nonce ('nonce-...') is required to enable inline execution.


I think the easiest thing to do here is to add a dedicated JS route which we can reference via `<script>` tag. [We do something similar for the SSO auth dance](https://github.com/elastic/kibana/blob/af09fedaf2dca691c4d81b5a0f83b0f9d8179c53/x-pack/plugins/security/server/routes/authentication/oidc.ts#L22-L83) today in order to capture the user's requested kibana endpoint.

```ts
        const body = renderToStaticMarkup(
          <>
            <ResetSessionPage
              logoutUrl={logoutUrl}
              styleSheetPaths={styleSheetPaths}
              uiPublicUrl={uiPublicUrl}
            />
            <script src={`${basePath}/internal/security/reset_session_page.js`} />
          </>
        );
```


And then in a route handler:
```ts
  httpResources.register(
    {
      path: '/internal/security/reset_session_page.js',
      validate: false,
      options: { authRequired: false },
    },
    (context, request, response) => {
      return response.renderJs({
        body: `
          document.getElementById('goBackButton').onclick = function () { window.history.back() }
        `,
      });
    }
  );
```

Reply:
Fixed